Allow bind mounts

david-mears-2 opened the pull request:
The former `ConstellationMount` was renamed `ConstellationVolumeMount` to reflect its inherent assumption that the Docker mount it wraps is a [volume mount](https://docs.docker.com/engine/storage/volumes/).

`ConstellationBindMount` is introduced, which allows constellation users to create [bind mounts](https://docs.docker.com/engine/storage/bind-mounts) that can be passed into `ConstellationContainer` by the 'mounts' parameter. `ConstellationContainer` requires/expects members of that parameter to implement a `to_mount(self, volumes)` interface.

The two `...Mount` classes are made to share a 'to_mount' interface so that 'mounts', the constructor parameter of `ConstellationContainer`, can be iterated over cleanly.

`ConstellationVolumeMount` now dictates the mount type as 'volume', reflecting its name; previously, 'volume' could have been overridden with another type via kwargs (although this would have caused an error when to_mount was called).

Version number updated as this would be a breaking change for any user who uses `ConstellationMount` - they only need to rename this import to `ConstellationVolumeMount`.

david-mears-2 committed:
Introduce ConstellationBindMount

The former ConstellationMount was renamed ConstellationVolumeMount to reflect its inherent
assumption that the mount it wraps is a volume mount.

ConstellationBindMount is introduced, which allows constellation users to create bind
mounts that can be passed into ConstellationContainer without requiring irrelevant
volume details to be passed to the ConstellationBindMount.

The interfaces of the two ...Mount classes are made identical so that 'mounts', the
constructor parameter of ConstellationContainer, can be iterated over cleanly.

ConstellationVolumeMount now dictates the mount type as 'volume', to reflect its name;

previously, 'volume' could have been overridden with another type via kwargs (although
this would have caused an error when to_mount was called).

david-mears-2 committed:
Have CVolumeMount and CBindMount inherit shared interface

david-mears-2 committed:
Format

david-mears-2 committed:
Remove pointless check

plietar commented on src/constellation/constellation.py:
```diff
@@ -335,18 +335,39 @@ def remove(self):
-        self.name = name
-        self.path = path
+# Base class for mounts.
+class _ConstellationMount:
```
Make this an ABC, or even better a Protocol (sort of like a Java interface): https://typing.python.org/en/latest/spec/protocol.html

I don’t really find that the common constructor adds that much value. I’d just make each class self contained that happen to implement the same structural interface.

plietar commented on src/constellation/constellation.py:
```diff
@@ -335,18 +335,39 @@ def remove(self):
+        self.target = target
         self.kwargs = kwargs
 
+    def to_mount(self, volumes):
```
Add `@abstractmethod`

plietar commented on src/constellation/constellation.py:
```diff
@@ -335,18 +335,39 @@ def remove(self):
 
+    def to_mount(self, volumes):
+        """Convert to a Docker mount. Subclasses can use `volumes` if needed."""
+        raise NotImplementedError("Subclasses must implement `to_mount`")
```
https://github.com/reside-ic/constellation/blob/master/pyproject.toml#L100

You can add `raise NotImplementedError` to that list to get coverage to shut up about these lines.

plietar replied:
https://coverage.readthedocs.io/en/latest/excluding.html

david-mears-2 committed:
Use Protocol instead of subclassing

david-mears-2 committed:
Format

david-mears-2 committed:
Silence CodeCov complaint

richfitz commented on src/constellation/__init__.py:
```diff
@@ -1,15 +1,17 @@
+    "ConstellationBindMount",
     "ConstellationContainer",
     "ConstellationService",
-    "ConstellationMount",
```
https://github.com/search?q=ConstellationMount&type=code - there are a few small changes we'll need to make:

* `bacpop/beebop-deploy`
* `vimc/montagu-deploy`
* `jameel-institute/daedalus-deploy`
* `mrc-ide/packit-deploy`
* `hivtools/hint-deploy`
* `vimc/orderly-web-deploy`

Each of these just needs version number requirements increasing and a change from `ConstellationMount` to `ConstellationVolumeMount`, which we can do as soon as this goes to PyPI

david-mears-2 replied:
I guess we should wait until this new version goes to PyPI before opening PRs on those packages.

richfitz commented on src/constellation/constellation.py:
```diff
@@ -335,18 +347,30 @@ def remove(self):
 
+class ConstellationBindMount:
+    def __init__(self, source, target, **kwargs):
+        self.source = source
```
The source here; can we do some processing on this. At the point where docker mounts the volume this needs to be an absolute path.

If provided as a relative path (`not Path(source).is_absolute()`) we should interpret it as relative to the root directory (I am fairly sure we have this concept).  Alternatively it could be interpreted relative to the working directory which might make the logic easier to understand, but will have the unfortunate consequence that where you run the command from will mount a different path.

Perhaps easiest at this point would be to require an absolute path, which is a bit gross in the configuration, but we can tidy that up in the library later - I don't think there's a good single place to get that information, and the whole package could do with a bit of TLC frankly.

Also check that `target` is an absolute path, that's at least easier...

david-mears-2 replied:
> The source here; can we do some processing on this. At the point where docker mounts the volume this needs to be an absolute path.

Since this processing will be needed for both types of mount, maybe I should go back to the subclassing pattern, _pace_ Paul, so that the parent class can implement a shared absolute-path-checking method.

> If provided as a relative path (not Path(source).is_absolute()) we should interpret it as relative to the root directory (I am fairly sure we have this concept). Alternatively it could be interpreted relative to the working directory which might make the logic easier to understand, but will have the unfortunate consequence that where you run the command from will mount a different path.

I don't see where constellation has the concept of a root directory; the constructor of `ConstellationContainer` does allow a `working_dir` to be configured.

richfitz replied:
I think you're right - it's not in there yet from the look of it, so we'll need to add that later. The `working_dir` argument for `ConstellationContainer` will be the path _within_ the container, so slightly different to what we want.

david-mears-2 committed:
Raise if mount target or source is not an absolute path

In the future we could interpret relative paths as relative to the root directory,
if we have that concept.

david-mears-2 commented:
> Perhaps easiest at this point would be to require an absolute path

This is the approach I've gone for now.

david-mears-2 committed:
Remove redundant exclusion from codecov

david-mears-2 committed:
Remove redundant import

david-mears-2 committed:
Remove type hints

david-mears-2 commented:
Codecov thinks I should have tested the tests?

richfitz commented:
> Codecov thinks I should have tested the tests?

This is alerting you to the file now having two functions with the same name, so one is never run

david-mears-2 committed:
Fix duplicate test name
